lib/statespace.h

```diff
@@ -94,9 +104,67 @@ class StateSpace {
   }
 
  protected:
-  uint64_t size_;
-  uint64_t raw_size_;
+  template <typename Func>
+  double Norm(uint64_t size, Func&& PartialNorms, const State& state) const {
+    auto partial_norms = PartialNorms(num_threads_, size, state);
+
+    double norm = partial_norms[0];
+    for (std::size_t i = 1; i < partial_norms.size(); ++i) {
+      norm += partial_norms[i];
+    }
+
+    return norm;
+  }
+
+  template <typename RGen, typename Func1, typename Func2>
+  MeasurementResult VirtualMeasure(
+      const std::vector<unsigned>& qubits, RGen& rgen, const State& state,
+      uint64_t size, Func1&& PartialNorms, Func2&& FindMeasuredBits) const {
+    MeasurementResult result;
+
+    result.valid = true;
+    result.mask = 0;
+
+    for (auto q : qubits) {
+      if (q >= num_qubits_) {
+        result.valid = false;
+        return result;
+      }
+
+      result.mask |= uint64_t{1} << q;
+    }
+
+    auto partial_norms = PartialNorms(num_threads_, size, state);
+
+    for (std::size_t i = 1; i < partial_norms.size(); ++i) {
+      partial_norms[i] += partial_norms[i - 1];
+    }
+
+    auto norm = partial_norms.back();
+    auto r = RandomValue(rgen, norm);
+
+    unsigned m = 0;
+    while (r > partial_norms[m]) ++m;
+    r -= m > 0 ? partial_norms[m - 1] : 0;
+
+    uint64_t k0 = For::GetIndex0(size, num_threads_, m);
+    uint64_t k1 = For::GetIndex1(size, num_threads_, m);
+
+    result.bits = FindMeasuredBits(k0, k1, r, result.mask, state);
```

Comment:
The way this uses `For` indices and the structure of the `FindMeasuredBits` method itself give me the sense that this ought to use `For::Run` - could you comment on this?

Reply:
No, we don't use `For::Run` here. In principle, we can repeat the previous step a few times recursively before that but this is probably beyond this PR. There is a for loop in `FindMeasuredBits` that need to be run sequentially. It begins with `k0` and ends with `k1`. Those `k0` and `k1` should be the same indices that were used by thread number `m` in `For::RunReduce` at the previous step, see above.